### ISSUE TITLE
Exit Kotekan on bad config update

### ISSUE THREAD

**Issue by cubranic** [kotekan/kotekan#450]
If the POSTed config is invalid, Kotekan will stop the mode and unwind back to a "not running" state, but this appears not to be 100% clean. E.g., this log shows that there were Stage objects still left after config update backed out, causing the system to fail on the next config update.

```
Jul 17 04:14:05 cn3g0 kotekan[5668]: Starting new kotekan mode using POSTed config.
... [ various stages starting up ]
Jul 17 04:14:41 cn3g0 kotekan[5668]: Failure reading 'enabled' from update: [json.exception.type_error.302] type must be boolean, but is string
Jul 17 04:14:41 cn3g0 kotekan[5668]: Runtime error configUpdater: Failure when calling subscriber to set initial value at endpoint '/updatable_config/gating/psr0_config'.
Jul 17 04:14:41 cn3g0 kotekan[5668]: configUpdater: Removing endpoint /frb_gain
Jul 17 04:14:41 cn3g0 kotekan[5668]: configUpdater: Removing endpoint /pulsar_gain
Jul 17 04:14:41 cn3g0 kotekan[5668]: configUpdater: Removing endpoint /updatable_config/gating/psr0_config
Jul 17 04:20:08 cn3g0 kotekan[5668]: Starting new kotekan mode using POSTed config.
.... [ various stages starting up ]
Jul 17 04:20:41 cn3g0 kotekan[5668]: Duplicate metric name: kotekan_baseband_readout_total:/baseband/baseband0. Exiting.
```

Instead, would it be better for Kotekan to just exit the process immediately on `configUpdater` failures, and so guarantee starting with a clean slate sooner? @andrerenard @nritsche @jrs65?

**Comment by nritsche** [comment]
I would propose to do that on the initial update (uses values from the config file) instead on all updates.

**Comment by andrerenard** [comment]
@nritsche I agree we should only exit on the initial update, and I think this is only an issue with the `/start` endpoint anyway.    The original idea was that we'd cleanly "unroll" a config state, and the stages _should_ all cleanly exit if we got a bad config.   However the ROCm (GPU) and DPDK (NIC) stages use hardware calls that don't have working close calls due to issues in those libraries; they expect a full binary shutdown.   So for that reason I'll change the behaviour here to exit the binary.   But I would still support having all other stages be able to shut down fully without a binary shutdown, because I'm hoping long term both ROCm and DPDK will add clean exit options.

**Comment by cubranic** [comment]
Does updating the config always start a new mode? Because Stages don't remove their metrics, `~kotekanMode` does.

**Comment by andrerenard** [comment]
A full new config generates a new mode, and shutting down deletes the mode.   An updatable config update doesn't change the mode or stages, it just adjusts the state of the stages.   In most cases an updatable config failure will simply reject the endpoint update, and in some cases _can_ trigger a shutdown.    However at the start of a new mode, the updatable config system runs on all stages once to provide the initial state.    It appears that this bug is happening in that initial application of the config state.

**Comment by andrerenard** [comment]
Implemented in #465
